### ISSUE TITLE
Arbitrarily oriented clipping planes capped with textures

### ISSUE THREAD

**Issue by nantille** [brainglobe/brainrender#127]
Hello, thanks for the interesting work that you published. Looking at using it, are multiple clipping planes implemented and if yes, is it possible to put a cap on the clipped meshes so that we see the internal structure of the brain per region or per custom data, applied as a texture on the cap(s) ?

Here's an example from XTK WebGL viewer with three clipping planes to illustrate what I mean.
![xtk_labelled_volume](https://user-images.githubusercontent.com/10866779/100124988-9248d080-2e7c-11eb-8e97-4d008385ad2f.png)

**Comment by FedeClaudi** [comment]
Hi, thanks for getting in touch. 

If I understand correctly, you have a 3-dimensional image with some data, right?
You can't currently render 3d volumes as they are in brainrender, you need to create a mesh from them. 
This is done automatically in brainrender using the [Volume](https://github.com/brainglobe/brainrender/blob/master/brainrender/actors/volume.py) actor class.
It looks something like:

<img width="1216" alt="image" src="https://user-images.githubusercontent.com/17436313/100131210-a9d38980-2e7b-11eb-927d-395c4fdf90f4.png">

here I have a 3d image and the value at each voxel shows the intensity of gene expression in that brain region, this is shown as 'cubes' in the rendering. Not all voxels are shown, only those with a gene expression level above a given threshold.
This is done using vedo's [Volume](https://github.com/marcomusy/vedo/blob/2e1c75bee15a3aa8ce58e3dfda16c615e7cf114b/vedo/volume.py#L231) class which supports other methods for creating meshes from images. 

Now, once you have created a mesh from your volumetric image, you can 'slice' the image using `Scene.slice` as you wish (with 3 orthogonal planes or any arbitrary plane), there's examples showing how to slice scenes.

I'm not sure if this addressed your question, if not perhaps you can tell me more about what data you have, what they look like and what you're aiming to achieve. One thing that is not clear from your question is if your data are registered to a reference atlas (and if so, which one?).

**Comment by FedeClaudi** [comment]
Sorry just one addition, from your image it looks like different portions of your 3d image are labelled differently, if so you can create a mesh from each which might look rather nice.

**Comment by nantille** [comment]
Thanks for your fast reply. The idea is to be able to move the slicing plane, revealing certain data. Here is the live demo from XTK samples : http://lessons.goxtk.com/11/. The only issue with such web frameworks is that 8y old code for web standard is like century old stuff unfortunately.

With what you say, I could generate a volume with custom data but if I want this to be interactive (if possible), then I have to recompute/redraw the volume, especially if it's clipped by multiple planes (like left, right, top, bottom and so on, if we imagine a box). Interesting anyway, I'll have a look at this.

When you say that brainrender cannot render volume, does it mean you take volumetric data with Volume class and do a marching cube on it so that you only display meshes?

**Comment by FedeClaudi** [comment]
Hi, 

I see, that clarifies things thanks. 

Some clarifications:
* vedo, which brainrender uses to handle the renderings, *can* handle volumes and does so beautifully. Check the exaples at [vedo](https://github.com/marcomusy/vedo), it can render volumes or extract density maps or surface meshes.
* currently brainrender does not work with Volumes however, only with meshes. So brainrender uses vedo's Volume class methods to create meshes from volumes. This can be done with a few different algorithms, not sure if marching cubes is one of these.
* If you want to use volumes, and not just meshes, I should be able to make ti happen. 

Now, as you say the more annoying part might be coding up the interactive slicing of the scene. However fear not, vedo has got this covered too:
![image](https://user-images.githubusercontent.com/17436313/100151570-393b6580-2e99-11eb-9cee-6470537ecf3f.png)

vedo has a `Slicer` widget that allows you to do exactly what you need ([example code](https://github.com/marcomusy/vedo/blob/c2cb3a4128dae6b7ddd144f31792d16080e4558e/examples/advanced/slicer1.py)). 

In conclusion, there's two options here as far as I can see:

1. if your data are **not** registered to a reference atlas supported by Brainglobe's AtlasAPI and you are not planning to register it an any point, your best option might be to just use `vedo` and its great `slicer` widget.
2. if, however, you do want to use your data together with some atlas (which one?), then you need brainrender. In this case I can make Volumes work in brainrender, and you can use vedo's Slicer as a starting point to code something similar using brainrender

**Comment by FedeClaudi** [comment]
Actually scratch that, I went and fixed brainrender so that it may work with volumes:
<img width="1680" alt="image" src="https://user-images.githubusercontent.com/17436313/100153688-5cb3df80-2e9c-11eb-9a30-97b06b01fcfc.png">

you may now use either vedo's slicer or implement something similar in brainrender, as you prefer ;)

**Comment by nantille** [comment]
Great news, thank you for the pointers to vedo, volume stuff, slicers and for the example code.
Actually we're using Allen CCF v3 as a basis and building from there (it looks like the hippocampi and mouse brain come from that atlas).
About this last image, you say it now works with volumes but I see a surface mesh (hippocampal formations). If you have a small code sample to show what you're doing here, that's welcome too. :)

**Comment by marcomusy** [comment]
it's not a surface, it's a volume visualised in "composite" mode.

**Comment by FedeClaudi** [comment]
Oh great news, if you're using the allen atlas then brainrender is perfect for you. 

It **looks** like a mesh, but it's really a volume (if you were to slice and look inside you'd see that it's full, unlike a mesh).
The reason why you're only seeing a shape that looks like the hippocampus is because the volume is thresholded to show only the voxels with higher gene expression (and interpolated to looks smooth).

I've added an example to show how to use volumes in brainrender: https://github.com/brainglobe/brainrender/blob/master/examples/volumetric_data.py

Note that brainrender's `Volume` class uses the `Volume` class from vedo to handle the rendering (I know, repeated names makes things confusing...). Vedo's `Volume` has many [options](https://github.com/marcomusy/vedo/blob/2e1c75bee15a3aa8ce58e3dfda16c615e7cf114b/vedo/volume.py#L231) to change how the volume should be rendered. You can pass keyword arguments to brainrender's `Volume`, these are passed. to vedo's `Volume` and can therefore affect how the volume is rendered. 

e.g. if I wanted to use `mode=2` in vedo's `Volume` I would do this:
``` python
actor = Volume(
    data,
    voxel_size=200,  # size of a voxel's edge in microns
    as_surface=False,  # if true a surface mesh is rendered instead of a volume
    c="Reds",  # use matplotlib colormaps to color the volume
    mode=2,  # <- get's passed to vedo
)
```

Have a go at playing around with the volumes visualizations in brainrender and let me know if you bump in any issue or have questions!

**Comment by nantille** [comment]
Thank you very much for your explanations Federico!
